### PR TITLE
feat: Adding volume mount for goal commands

### PR DESCRIPTION
**Pull request by negar-abbasi**

### CODE REVIEW

File: tests/goal/test_goal.py [hunk 1]
```diff
@@ -80,6 +80,30 @@ def test_goal_simple_args() -> None:
...
+def test_goal_simple_args_with_input_file() -> None:
+    result = invoke("goal clerk compile approval.teal")
+
+    assert result.exit_code == 0
+    verify(result.output)
+
+
+@pytest.mark.usefixtures("proc_mock")
+def test_goal_simple_args_with_output_file() -> None:
+    result = invoke("goal clerk compile approval.compiled")
```
Comment (aorumbayev) [inline tests/goal/test_goal.py:93]: Are you sure this is `with output file` ? I think you need an explicit flag, also compile command migh always expect an input to be provided otherwise it might not know what to compile

File: tests/goal/test_goal.py [hunk 2]
```diff
@@ -80,6 +80,30 @@ def test_goal_simple_args() -> None:
...
+
+    assert result.exit_code == 0
+    verify(result.output)
+
+
+@pytest.mark.usefixtures("proc_mock")
+def test_goal_simple_args_with_input_output_files() -> None:
+    result = invoke("goal clerk compile approval.teal approval.compiled")
```
Comment (aorumbayev) [inline tests/goal/test_goal.py:101]: when you don't provide -o flag it assumes all inputs after compile are batched inputs, make sure to test cases when you have several input files and a single output as well.

File: tests/goal/test_goal.py [hunk 3]
```diff
@@ -80,6 +80,30 @@ def test_goal_simple_args() -> None:
     verify(result.output)
 
 
+@pytest.mark.usefixtures("proc_mock")
+def test_goal_simple_args_with_input_file() -> None:
+    result = invoke("goal clerk compile approval.teal")
```
Comment (aorumbayev) [inline tests/goal/test_goal.py:85]: with proc_mock you are probably mocking the execution of the docker itself but remaining bits arent mocked (which makes sense). So id suggest to have a real simple text file (doesn't even have to be teal) but it needs to exist on the cwd path if you want to test:
- paths params being changed by pre processing and post processing
- input file being copied to and from `mocked` volume mount folder. You can mock the `out` file to be returned back to return some other dummy text file (since you arent gonna run actual goal clerk)

File: tests/goal/test_goal.py [hunk 4]
```diff
@@ -80,6 +80,30 @@ def test_goal_simple_args() -> None:
     verify(result.output)
 
 
+@pytest.mark.usefixtures("proc_mock")
+def test_goal_simple_args_with_input_file() -> None:
```
Comment (aorumbayev) [inline tests/goal/test_goal.py:84]: aside from this, make sure to check on other goal command that may require dealing with files, if you see any interesting edge cases or alternative uses (for instance whether there are any command that allow for N input and N outputs) make sure to cover such edge cases too.

File: src/algokit/core/goal.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+from algokit.core.conf import get_app_config_dir
+
+logger = logging.getLogger(__name__)
+
+
+def get_volume_mount_path_docker() -> Path:
+    os_type = platform.system().lower()
+    if os_type == "windows":
+        return Path(r"\root\goal_mount")
```
Comment (robdmoore) [inline src/algokit/core/goal.py:15]: Is this right? What would this do?

File: src/algokit/cli/goal.py [hunk 6]
```diff
@@ -44,13 +53,19 @@ def goal_command(*, console: bool, goal_args: list[str]) -> None:
...
+        cmd = cmd + goal_args
         result = proc.run(
             cmd,
             stdout_log_level=logging.INFO,
             prefix_process=False,
             pass_stdin=True,
         )
+        post_process(input_files, output_files, volume_mount_path_local)
+
+    # TODO: Update this block to provide proper warnings for file/folder not found
```
[the remark this answers is not in the document]
Comment (negar-abbasi) [reply src/algokit/cli/goal.py:74]: No it is already done. I'll remove the todo tags.

File: docs/features/goal.md [hunk 7]
```diff
@@ -78,13 +78,29 @@ $ ~ algokit goal
...
+When interacting with the container, especially if you're using tools like goal, you might need to reference files or directories. Here's how to efficiently deal with files and directories:
 
-If you want to copy files into the container so you can access them via goal then you can use the following:
+### Automatic File Mounting
+When you specify a file or directory path in your `goal` command, the system will automatically mount that path from your local filesystem into the container. This way, you don't need to copy files manually each time.
+
+For instance, if you want to compile a `teal` file:
+
+```
+goal clerk compile /Path/to/inputfile/approval.teal -o /Path/to/outputfile/approval.compiled
```
Comment (aorumbayev) [inline docs/features/goal.md:90]:
```suggestion
algokit goal clerk compile /Path/to/inputfile/approval.teal -o /Path/to/outputfile/approval.compiled
```

File: tests/utils/proc_mock.py [hunk 8]
```diff
@@ -88,8 +90,27 @@ def should_bad_exit_on(self, cmd: list[str] | str, exit_code: int = -1, output:
...
+    ) -> None:
+        """
+        Set the output of a command, and optionally a side effect to be called when the command is run.
+        Side effect can't have input arguments, the main utility is running extra logic on top of execution of the
+        proc_mock object. Common scenario includes, generating dummy files as a side effect of invoking the command.
+
+        Args:
+            cmd: The command to set the output for
+            output: The output to return when the command is run
+            side_effect: A callable to be called when the command is run
```
Comment (aorumbayev) [inline tests/utils/proc_mock.py:108]:
```suggestion
            side_effect: A callable to be called when the command is run
            side_effect_args: Key value paired arguments to pass to the side_effect function (optional)
```

File: tests/utils/proc_mock.py [hunk 9]
```diff
@@ -88,8 +90,27 @@ def should_bad_exit_on(self, cmd: list[str] | str, exit_code: int = -1, output:
...
+    def set_output(
+        self,
+        cmd: list[str] | str,
+        output: list[str],
+        side_effect: Callable[[Any], None] | None = None,
+        side_effect_args: dict[str, Any] | None = None,
+    ) -> None:
+        """
+        Set the output of a command, and optionally a side effect to be called when the command is run.
+        Side effect can't have input arguments, the main utility is running extra logic on top of execution of the
```
Comment (aorumbayev) [inline tests/utils/proc_mock.py:102]: Change the wording here, side effect can have input arguments now so probably just explain what side effect is (allows you to execute some custom function during execution of a mock proc_mock. But the sentenc saying it can't have input args should be removed as we pass the side_effect_args now

**Review by aorumbayev — APPROVED** [review]
@negar-abbasi minor tweaks left please address otherwise LGTM. 
Approving the pr, @robdmoore if you'll have time to take a peek on Monday for final sanity check would be great, will try to coordinate to have this as part of 1.4.1 release on monday

**Review by neilcampbell — APPROVED** [review]
